src/sampling.jl

```diff
@@ -42,6 +42,47 @@ function direct_sample!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
 end
 direct_sample!(a::AbstractArray, x::AbstractArray) = direct_sample!(Random.GLOBAL_RNG, a, x)
 
+# check whether we can use T to store indices 1:n exactly
+for T in unique!([Int,UInt,Int64,UInt64,Int128,UInt128])
+    @eval isexact(n, ::Type{$T}) = true
+end
+isexact(n, ::Type{T}) where {T<:Integer} = n ≤ typemax(T)
+isexact(n, ::Type{BigInt}) = false # bigint is too expensive
+for T in (Float32, Float64)
+    @eval isexact(n, ::Type{$T}) = n ≤ maxintfloat($T)
+end
+isexact(n, ::Type{Complex{T}}) where {T} = isexact(n, T)
+isexact(n, ::Type{Rational{T}}) where {T} = isexact(n, T)
+isexact(n, T) = false
+
+# order results of a sampler that does not order automatically
+function sample_ordered!(sampler!::F, rng::AbstractRNG, a::AbstractArray, x::AbstractArray) where {F}
```

Comment:
Why do you force specialization on `F`? This happens automatically AFAIK.

Reply:
It doesn't always happen automatically for function arguments: https://docs.julialang.org/en/v1/manual/performance-tips/#Be-aware-of-when-Julia-avoids-specializing

(Of course, the specialization probably doesn't matter here since the inner loops occur inside the `sampler!` function.)

Reply:
"Julia will always specialize when the argument is used within the method, but not if the argument is just passed through to another function." So I think we're safe. Actually I didn't know it was documented so clearly. :-)

Reply:
Done.